contracts/axelarnet-gateway/src/contract/execute.rs

```diff
@@ -79,297 +180,37 @@ pub fn receive_messages(
     ))
 }
 
-pub fn send_messages(
+/// Route messages to the router, ignore unknown messages.
+fn route_to_router(
     store: &mut dyn Storage,
     router: &Router,
     msgs: Vec<Message>,
 ) -> Result<Response, Error> {
-    for msg in msgs.iter() {
-        let stored_msg = state::may_load_sent_msg(store, &msg.cc_id)
-            .change_context(Error::InvalidStoreAccess)?;
-
-        match stored_msg {
-            Some(message) if msg != &message => {
-                Err(report!(Error::MessageMismatch(msg.cc_id.clone())))
-            }
-            Some(_) => Ok(()),
-            None => Err(report!(Error::MessageNotFound(msg.cc_id.clone()))),
-        }?
-    }
-
-    let (wasm_msg, events) = route(router, msgs)?;
-
-    Ok(Response::new().add_message(wasm_msg).add_events(events))
-}
-
-fn route(
-    router: &Router,
-    msgs: Vec<Message>,
-) -> Result<(WasmMsg, impl IntoIterator<Item = Event>), Error> {
-    Ok((
-        router.route(msgs.clone()).ok_or(Error::RoutingFailed)?,
-        msgs.into_iter()
-            .map(|msg| AxelarnetGatewayEvent::Routing { msg }.into()),
-    ))
-}
-
-pub fn execute(
-    store: &mut dyn Storage,
-    api: &dyn Api,
-    querier: QuerierWrapper,
-    cc_id: CrossChainId,
-    payload: HexBinary,
-) -> Result<Response, Error> {
-    let msg = state::set_msg_as_executed(store, cc_id.clone())
-        .change_context(Error::SetMessageStatusExecutedFailed(cc_id))?;
+    let msgs: Vec<_> = msgs
+        .into_iter()
+        .unique()
+        .filter_map(|msg| verify_message(store, msg).transpose())
```

Comment:
need to see if return value from verify can be avoided. otherwise may be rename it to something that implies a return value is being used

Reply:
I don't get it

Reply:
Sorry, referring to the `verify_message` returning a Message discussion

Reply:
How about creating a `try_filter` similar to the `try_map` that we have that handles this internally. Then the usage here is more clear, and `verify_message` just verifies instead of transforming the value.
`.try_filter(|msg| verify_message(store, &msg))?;`